agent/command.go

```diff
@@ -215,6 +215,11 @@ func (a *Agent) runCommandOrFunc(ctx context.Context, tc *taskContext, commandIn
 			commandSpan.SetStatus(codes.Error, "running command")
 			commandSpan.RecordError(err, trace.WithAttributes(tc.taskConfig.TaskAttributes()...))
 			commandSpan.End()
+			if commandInfo.AutoRetry {
+				if err2 := a.comm.MarkTaskToRestart(ctx, tc.task); err2 != nil {
```

Comment:
Should the command still auto-restart if it can't fail the task? Let's say a shell.exec fails, but it's in one of those commands that can't fail the task (say it fails in the pre block, but the YAML has set `pre_error_fails_task: false`). The task then continues on happily and ends in success, but still restarts at the end of teh task, even though the command's failure had no bearing on the task's status.

I guess if I were to phrase the question, is auto-restart meant to kick in when a command fails, or when a command fails and will subsequently cause the task to fail?

Reply:
> is auto-restart meant to kick in when a command fails, or when a command fails and will subsequently cause the task to fail?
My intention is the latter. In the case that `pre_error_fails_task` is set, the `runCommand` actually will not fail because this [conditional](https://gist.github.com/hadjri/e2c7bb1b22b261488d0c4737058a6c6c#file-command-go-L304) won't execute (which is derived from PreErrorFailsTask)

Reply:
That makes sense, but [this test](https://github.com/evergreen-ci/evergreen/blob/1e9f3acd58c48176a5d4ce1fcd1fe30066f5b864/agent/agent_test.go#L823) (which is implicitly using the default of `pre_error_fails_task: false`) seems like it's expecting (and actually sees) the opposite, because it's claiming that the task will auto-restart even though the pre command is not one that can fail the task.

It seems like you would have to check `canFailTask` explicitly before marking the task for restart, no? `canFailTask` only returns on error at the block level, not at the individual command level.